File: 2026-bogota-agent-sdk/typescript/03_tool.ts
// 03 · Una tool propia. tool() describe la función, createSdkMcpServer la publica.
// El nombre que ve el agente es mcp__{servidor}__{tool}: aquí mcp__reloj__hoy.
import { createSdkMcpServer, query, tool } from "@anthropic-ai/claude-agent-sdk";

// El tercer argumento es el esquema de entrada en zod. Aquí no recibe nada.
const hoy = tool(
  "hoy",
  "Devuelve la fecha de hoy en español.",
  {},
  async () => ({
    content: [{ type: "text" as const, text: "sábado 19 de septiembre de 2026" }],
  }),
);

// alwaysLoad evita el paso previo de ToolSearch: el agente ve la tool completa desde el init.
const reloj = createSdkMcpServer({ name: "reloj", tools: [hoy], alwaysLoad: true });

const conversacion = query({
  prompt: "¿Qué fecha es hoy y qué día de la semana cae?",
  options: {
    model: "claude-sonnet-5",
    systemPrompt:
      "Eres un asistente personal. Respondes en español, corto y con datos. " +
      "Usa tus tools y skills; no adivines fechas ni reservas.",

    // tools decide QUÉ EXISTE: [] apaga todas las tools de Claude Code
    // (Read, Bash, Edit...). Las del servidor MCP siguen existiendo.
    tools: [],
    mcpServers: { reloj },

    // allowedTools decide QUÉ SE APRUEBA SIN PREGUNTAR. Sin esta línea el
    // agente pediría permiso y, sin nadie que responda, la llamada se cae.
    allowedTools: ["mcp__reloj__hoy"],

    settingSources: [],
    strictMcpConfig: true,
    maxTurns: 4,
    maxBudgetUsd: 0.05,
    env: { ...process.env, CLAUDE_CODE_DISABLE_AUTO_MEMORY: "1" },
  },
});

for await (const m of conversacion) {
  if (m.type === "assistant") {
    for (const b of m.message.content) {
      if (b.type === "tool_use") console.log(`-> llama ${b.name}`);
      if (b.type === "text" && b.text.trim()) console.log(`Agente: ${b.text.trim()}`);
    }
  }
  if (m.type === "result" && m.subtype === "success") {
    console.log(`\nCosto: $${m.total_cost_usd.toFixed(4)}`);
  }
}

File: 2026-bogota-agent-sdk/python/03_tool.py
"""03 · una tool propia — `hoy()`, una función tuya que el LLM puede llamar.

Tres piezas:
  @tool(nombre, descripción, esquema)   la descripción es lo que el LLM lee
                                        para decidir si la llama. Escríbela bien.
  create_sdk_mcp_server(...)            las empaqueta en un servidor MCP que vive
                                        dentro de este mismo proceso de Python
  mcp_servers={"reloj": RELOJ}          la clave "reloj" da el nombre completo
                                        de la tool: mcp__reloj__hoy

Y dos listas que se confunden todo el tiempo:
  tools          qué existe          (`[]` = ninguna nativa; el set base)
  allowed_tools  qué se aprueba solo (no restringe: solo evita la pregunta)

La fecha va fija a propósito: el mismo resultado hoy y el día del taller.

    uv run python 03_tool.py
"""

import asyncio

from claude_agent_sdk import ClaudeAgentOptions, create_sdk_mcp_server, query, tool

from mostrar import linea

PREGUNTA = "Revisa mis correos de la última semana y dime cuándo es mi próximo vuelo."

SYSTEM_PROMPT = """Eres un asistente personal. Respondes en español, corto y con datos.
Usa tus tools y skills; no adivines fechas ni reservas."""


@tool("hoy", "La fecha de hoy. Pídela siempre antes de razonar sobre fechas.", {})
async def hoy(args):
    print("      ⚙  corriendo hoy() en tu Python")
    return {"content": [{"type": "text", "text": "sábado 19 de septiembre de 2026"}]}


RELOJ = create_sdk_mcp_server(name="reloj", version="1.0.0", tools=[hoy])


async def main():
    opciones = ClaudeAgentOptions(
        model="claude-sonnet-5",
        system_prompt=SYSTEM_PROMPT,
        tools=[],
        mcp_servers={"reloj": RELOJ},
        allowed_tools=["mcp__reloj__hoy"],   # quítalo y la corrida se queda esperando permiso
        setting_sources=[],
        strict_mcp_config=True,
        env={"CLAUDE_CODE_DISABLE_AUTO_MEMORY": "1"},
        max_turns=4,
        max_budget_usd=0.10,
        stderr=lambda _: None,
    )

    print(f"PREGUNTA  {PREGUNTA}\n")
    async for m in query(prompt=PREGUNTA, options=opciones):
        linea(m)


asyncio.run(main())
